File: 733. Flood Fill/map.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original = image[sr][sc]
        self.dfs(image, sr, sc, original, color)
        return image

    def dfs(self, image: List[List[int]], i: int, j: int, original: int, update: int):
        if i < 0 or i >= len(image) or j < 0 or j >= len(image[0]) or image[i][j] != original or image[i][j] == update:
            return
        image[i][j] = update
        self.dfs(image, i - 1, j, original, update)
        self.dfs(image, i + 1, j, original, update)
        self.dfs(image, i, j - 1, original, update)
        self.dfs(image, i, j + 1, original, update)

File: 733. Flood Fill/map.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int original = image[sr][sc];
        dfs(image, sr, sc, original, color);
        return image;
    }

    void dfs(vector<vector<int>>& img, int i, int j, int original, int update) {
        if (i < 0 || i >= img.size() || j < 0 || j >= img[0].size() || original != img[i][j] || img[i][j] == update) {
            return;
        }
        img[i][j] = update;
        dfs(img, i , j + 1, original, update);
        dfs(img, i , j - 1, original, update);
        dfs(img, i - 1, j, original, update);
        dfs(img, i + 1, j, original, update);
    }
};